share/shwap/p2p/shrex/server.go

```diff
@@ -0,0 +1,239 @@
+		srv.SetHandler(ProtocolID(srv.params.NetworkID(), protocolName), withRecovery)
+	}
+	return srv, nil
+}
+
+func (srv *Server) SetHandler(p protocol.ID, h network.StreamHandler) {
+	srv.host.SetStreamHandler(p, h)
+}
+
+// Stop stops the Server
+func (srv *Server) Stop(_ context.Context) error {
+	srv.cancel()
+	for _, id := range srv.supportedProtocols {
+		srv.host.RemoveStreamHandler(ProtocolID(srv.params.NetworkID(), id))
+	}
+	return nil
+}
+
+func (srv *Server) streamHandler(ctx context.Context, id newID) network.StreamHandler {
+	return func(s network.Stream) {
+		if !srv.handleDataRequest(ctx, id, s) {
+			s.Reset() //nolint:errcheck
+		}
+		if err := s.Close(); err != nil {
+			log.Debugw("server: closing stream", "err", err)
+		}
+	}
+}
+
+func (srv *Server) handleDataRequest(ctx context.Context, id newID, stream network.Stream) bool {
+	requestID := id()
+	logger := log.With("source", "Server", "name", "peer", requestID.Name(), stream.Conn().RemotePeer().String())
+	logger.Debug("handling data request")
+
+	// registering handlers are done only after the verification that
+	// protocol supports it. There is no need to additionally verify here whether we support it
+	// or not.
+
+	err := srv.readRequest(logger, requestID, stream)
+	if err != nil {
+		logger.Warnw("read request", "err", err)
+		srv.metrics.observeRequests(ctx, 1, requestID.Name(), StatusBadRequest)
+		return false
+	}
+
+	err = requestID.Validate()
+	if err != nil {
+		logger.Warnw("validate request", "err", err)
+		srv.metrics.observeRequests(ctx, 1, requestID.Name(), StatusBadRequest)
+		return false
+	}
+
+	logger.Debugw("new request")
+	ctx, cancel := context.WithTimeout(ctx, srv.params.HandleRequestTimeout)
+	defer cancel()
+
+	r, status, err := srv.getData(ctx, requestID)
+	sendErr := srv.respondStatus(ctx, logger, requestID, stream, status)
+	if sendErr != nil {
+		logger.Errorw("sending response status", "err", sendErr)
+		srv.metrics.observeRequests(ctx, 1, requestID.Name(), StatusSendRespErr)
+	}
+
+	if err != nil {
+		logger.Errorw("handling request", "err", errors.Join(err, sendErr))
+		return false
+	}
+
+	if status != shrexpb.Status_OK {
+		return false
+	}
+
+	_, err = io.Copy(stream, r)
+	if err != nil {
+		logger.Errorw("send data", "err", err)
+		srv.metrics.observeRequests(ctx, 1, requestID.Name(), StatusSendRespErr)
+		return false
+	}
+	return true
+}
+
+func (srv *Server) readRequest(
+	logger *zap.SugaredLogger,
+	id id,
+	stream network.Stream,
+) error {
+	err := stream.SetReadDeadline(time.Now().Add(srv.params.ServerReadTimeout))
+	if err != nil {
+		logger.Debugw("setting read deadline", "err", err)
+	}
+
+	_, err = id.ReadFrom(stream)
+	if err != nil {
+		return fmt.Errorf("reading request: %w", err)
+	}
+
+	err = stream.CloseRead()
+	if err != nil {
+		logger.Warnw("closing read side of the stream", "err", err)
+	}
+
+	return nil
+}
+
+func (srv *Server) getData(
+	ctx context.Context,
+	id id,
+) (io.Reader, shrexpb.Status, error) {
```

Comment:
Imo we should remove this func. It makes it complicated to track down what error is returned and what is observed